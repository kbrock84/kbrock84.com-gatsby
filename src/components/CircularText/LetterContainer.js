import styled from "styled-components";

const LetterContainer = styled.div`
  position: absolute;
  margin: 0;
  padding: 0;
  width: 8px;
  line-height: ${props => props.fontSize};
  vertical-align: text-bottom;
  font-family: monospace;
  font-size: ${props => props.fontSize};
  display: flex;
  flex-direction: column-reverse;
  align-items: center;
  height: ${props => props.r}px;
  color: ${props => props.color};
  transform-origin: top center;
`;

export default LetterContainer;
