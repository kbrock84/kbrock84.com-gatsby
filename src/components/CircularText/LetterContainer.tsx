import styled from "styled-components";

const LetterContainer = styled.div<{
  readonly fontSize: string;
  readonly r: number;
}>`
  position: absolute;
  margin: 0;
  padding: 0;
  width: 8px;
  line-height: ${props => props.fontSize};
  vertical-align: text-bottom;
  font-family: monospace;
  font-size: ${props => props.fontSize};
  display: flex;
  flex-direction: column-reverse;
  align-items: center;
  height: ${props => props.r}px;
  transform-origin: top center;
`;

export default LetterContainer;
