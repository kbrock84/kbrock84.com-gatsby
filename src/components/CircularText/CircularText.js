import React from "react";
import LetterContainer from "./LetterContainer";

const CircularText = props => {
  return (
    <div
      style={{
        color: "#ff00ff",
        display: "flex",
        flexDirection: "row",
        justifyContent: "center",
        alignItems: "center"
      }}
    >
      <div
        style={{
          position: "relative",
          height: `${props.r * 2}px`,
          width: `${props.r * 2}px`
        }}
      >
        {props.text.split("").map((char, index) => (
          <LetterContainer
            key={`circular-text-${index}`}
            r={props.r}
            style={{
              transform: `
               translate(${props.r - 4}px, ${props.r}px)
                rotate(${2.5 * index * -1}deg)`
            }}
          >
            {char}
          </LetterContainer>
        ))}
        {props.children}
      </div>
    </div>
  );
};

export default CircularText;
