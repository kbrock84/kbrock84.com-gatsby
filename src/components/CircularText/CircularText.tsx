import * as React from "react";
import LetterContainer from "./LetterContainer";

const CircularText: React.FunctionComponent<{
  r: number;
  text: string;
  offset?: number;
  fontSize?: string;
}> = props => {
  const offset: number = props.offset || 2.5;
  const fontSize: string = props.fontSize || "1em";
  return (
    <div
      style={{
        color: "#ff00ff",
        display: "flex",
        flexDirection: "row",
        justifyContent: "center",
        alignItems: "center"
      }}
    >
      <div
        style={{
          position: "relative",
          height: `${props.r * 2}px`,
          width: `${props.r * 2}px`
        }}
        data-testid="test-letter-container"
      >
        {props.text.split("").map((char, index) => (
          <LetterContainer
            key={`circular-text-${index}`}
            r={props.r}
            style={{
              transform:
                `translate(${props.r - 4}px, ${props.r}px) ` +
                `rotate(${offset * index * -1}deg)`
            }}
            fontSize={fontSize}
          >
            {char}
          </LetterContainer>
        ))}
        {props.children}
      </div>
    </div>
  );
};

export default CircularText;
