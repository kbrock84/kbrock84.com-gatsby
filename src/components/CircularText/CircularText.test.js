import React from "react";
import {
  render,
  cleanup,
  queryAllByLabelText,
  waitForElement
} from "react-testing-library";
import "jest-dom/extend-expect";

import CircularText from "./CircularText.js";
describe("<CircularText />", () => {
  it("Matches the previous accepted snapshot.", () => {
    const { asFragment } = render(
      <CircularText r="100" text="render this in a circle" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  cleanup();

  it("Displays an array of divs with each char passed to the text prop", () => {
    const text = "render this in a circle";
    const { getByTestId } = render(
      <CircularText r="100" text="render this in a circle" />
    );

    const letterDivs = getByTestId("test-letter-container").querySelectorAll(
      "div"
    );

    expect(letterDivs.length).toBe(text.length);
    letterDivs.forEach((div, i) => {
      expect(div.innerHTML).toBe(text[i]);
    });
  });

  it("displays the divs with a height if prop r + px", () => {
    const text = "render this in a circle";
    const radius = 100;
    const { getByTestId, debug } = render(
      <CircularText r="100" text="render this in a circle" />
    );

    const container = getByTestId("test-letter-container");
    container.childNodes.forEach((node, i) => {
      const div = getComputedStyle(node);
      expect(div.height).toBe(radius + "px");
    });
  });

  it("rotates the div counter-clockwise using the offset prop", () => {
    const text = "render this in a circle";
    const radius = 100;
    const { getByTestId, debug } = render(
      <CircularText r="100" text="render this in a circle" offset="10" />
    );

    const container = getByTestId("test-letter-container");
    container.childNodes.forEach((node, i) => {
      const div = getComputedStyle(node);
      const pattern = new RegExp(`.*rotate\\(${i * 10 * -1}deg\\).*`);
      const isRotated = pattern.test(div.transform);
      expect(isRotated).toBe(true);
    });
  });
});
