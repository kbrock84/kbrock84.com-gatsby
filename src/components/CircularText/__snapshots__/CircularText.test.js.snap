// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircularText /> Matches the previous accepted snapshot. 1`] = `
<DocumentFragment>
  <div
    style="color: rgb(255, 0, 255); display: flex; flex-direction: row; justify-content: center; align-items: center;"
  >
    <div
      data-testid="test-letter-container"
      style="position: relative; height: 200px; width: 200px;"
    >
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        r
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        e
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        n
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        d
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        e
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        r
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
         
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        t
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        h
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        i
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        s
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
         
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        i
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        n
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
         
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        a
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
         
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        c
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        i
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        r
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        c
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        l
      </div>
      <div
        class="sc-bdVaJa klusQH"
        r="100"
        style="transform: translate(96px, 100px) rotate(NaNdeg);"
      >
        e
      </div>
    </div>
  </div>
</DocumentFragment>
`;
